### DISCUSSION TITLE
Template project structure

### DISCUSSION THREAD

**Discussion by mturoci** [h2oai/wave#673] [Ideas]
Problem: App developers usually use a single `.py` file for writing their apps. This results in 1500+ lines files that are messy and hard to debug.

Solution: It would be nice to provide app developers a recommended folder structure together with not only visual templates for their apps, but also things like recommended way of working with local assets (images), file uploads etc.

We could create a CLI (or make it part of existing q8s CLI maybe?). Another option would be to use [Github template repos](https://docs.github.com/en/free-pro-team@latest/github/creating-cloning-and-archiving-repositories/creating-a-template-repository) - this would however tightly couple the project to _Github_ as a VCS not being truly platform agnostic solution that we should go for. Users can of course change git remote origin, but this would assume certain git knowledge we probably don't want our users to be bothered by.

Let's use this issue as a starting point for discussion.

cc @geomodular @lo5

**Comment by lo5** [comment]
@mturoci Is there anything specific you're recommending or is #38 enough?

**Comment by mturoci** [comment]
I have noticed during one of the @mtanco 's demos that some people use Pycharm wizard for creating a new project. This could be done via our custom Pycharm plugin. I still consider cli a better option though, but might be just my personal preference.

#38 AFAIU deals with ui layouts, not project structure in general or maybe I am wrong?

Another place for recommending the best practices can be docs page and public OSS apps repo.
